src/app/3d/qgs3dmaptoolmeasureline.cpp

```diff
@@ -0,0 +1,269 @@
+  if ( QgsTerrainEntity *terrainEntity = mCanvas->scene()->terrainEntity() )
+  {
+    connect( terrainEntity->terrainPicker(), &Qt3DRender::QObjectPicker::clicked, this, &Qgs3DMapToolMeasureLine::onTerrainPicked );
+  }
+}
+
+void Qgs3DMapToolMeasureLine::handleClick( Qt3DRender::QPickEvent *event, const QgsVector3D &worldIntersection )
+{
+  if ( event->button() == Qt3DRender::QPickEvent::LeftButton )
+  {
+    if ( mDone )
+    {
+      mDialog->restart();
+    }
+
+    mDone = false;
+    QgsVector3D mapCoords = Qgs3DUtils::worldToMapCoordinates( QgsVector3D( worldIntersection.x(),
+                            worldIntersection.y(),
+                            worldIntersection.z() ), mCanvas->map()->origin() );
+    addPoint( QgsPoint( mapCoords.x(), mapCoords.y(), mapCoords.z() ) );
+    mDialog->show();
+  }
+  else if ( event->button() == Qt3DRender::QPickEvent::RightButton )
+  {
+    // Finish measurement
+    mDone = true;
+  }
+  else if ( event->button() == Qt3DRender::QPickEvent::MiddleButton )
+  {
+    undo();
+  }
+}
+
+void Qgs3DMapToolMeasureLine::setMeasurementLayerRenderer()
+{
+  mMeasurementLayer->setRenderer3D( mLineSymbolRenderer );
+  mLineSymbolRenderer->setLayer( mMeasurementLayer );
+}
+
+void Qgs3DMapToolMeasureLine::updateMeasurementLayer()
+{
+  double verticalScale = canvas()->map()->terrainVerticalScale();
+  QgsLineString *line;
+  if ( verticalScale != 1.0 )
+  {
+    QVector<QgsPoint> descaledPoints;
+    QVector<QgsPoint>::const_iterator it;
+    QgsPoint point;
+    for ( it = mPoints.constBegin(); it != mPoints.constEnd(); ++it )
+    {
+      point = *it;
+      descaledPoints.append(
+        QgsPoint( it->x(), it->y(), it->z() / verticalScale )
+      );
+    }
+    line = new QgsLineString( descaledPoints );
+  }
+  else
+  {
+    line = new QgsLineString( mPoints );
+  }
+  QgsGeometry *lineGeometry = new QgsGeometry( line );
+
+  QgsGeometryMap geometryMap;
+  geometryMap.insert( 1, *lineGeometry );
+  mMeasurementLayer->dataProvider()->changeGeometryValues( geometryMap );
+  mCanvas->map()->setRenderers( QList<QgsAbstract3DRenderer *>() << mMeasurementLayer->renderer3D()->clone() );
+}
+
+void Qgs3DMapToolMeasureLine::updateSettings()
+{
+  // Line style
+  QgsLine3DSymbol *lineSymbol = new QgsLine3DSymbol;
+  lineSymbol->setRenderAsSimpleLines( true );
+  lineSymbol->setWidth( 4 );
+  lineSymbol->setAltitudeClamping( Qgs3DTypes::AltClampAbsolute );
+
+  QgsPhongMaterialSettings phongMaterial;
+  QgsSettings settings;
+  int myRed = settings.value( QStringLiteral( "qgis/default_measure_color_red" ), 222 ).toInt();
+  int myGreen = settings.value( QStringLiteral( "qgis/default_measure_color_green" ), 155 ).toInt();
+  int myBlue = settings.value( QStringLiteral( "qgis/default_measure_color_blue" ), 67 ).toInt();
+  phongMaterial.setAmbient( QColor( myRed, myGreen, myBlue, 100 ) );
+  lineSymbol->setMaterial( phongMaterial );
+
+  mLineSymbolRenderer = new QgsVectorLayer3DRenderer( lineSymbol );
```

Comment:
At this point, a previous mLineSymbolRenderer object may still exist, so there may be a memory leak here. Best to use std::unique_ptr that takes care of destructing the old 3d renderer when needed.

Reply:
Hi @wonder-sk 
Actually, I pass that `mLineSymbolRenderer` to the `mMeasurementLayer` in the `setMeasurementLayerRenderer` function. It set by `mMeasurementLayer->setRenderer3D( lineSymbolRenderer );`, I check in the code, whenever it calls `setRenderer3D`, it will delete the old `3DRenderer`.

So, it's safe to use `new` here, isn't it?

Btw, if I use std:unique_ptr, it will crash whenever I call the `setMeasurementLayerRenderer`. Probably because it is pointed by the layer now.

Reply:
true - sorry my bad, I have forgotten the ownership is taken by the layer...